### PR TITLE
feat: root ca generation.

### LINKED ISSUE #544: Generate CA and certificate
- Generate Certificate Authority
- Generate certificate for `kubewarden-controller` and sign it with our CA
- Start the Kubewarden controller HTTPS server using this certificate
- Register the webhook against the Kubernetes API, using the certificate of our CA inside of
  `clientConfig.caBundle`
  

Related to: https://github.com/kubewarden/rfc/blob/main/rfc/0018-certificate-handling.md#kubewarden-fresh-install

### Alternatives you've considered

Maybe this could be helpful: https://github.com/open-policy-agent/cert-controller

### Anything else?

Remove cert manager dead code.

### PR DESCRIPTION
**Pull request by jvanz**
## Description

Adds code to create a root CA, create another certificates signed by the root CA create previously and use this certificate in the webhook server. As well as update the webhooks to use the root CA certificate to validate the webhook server certificates.

Fix #544 

**IMPORTANT NOTES BEFORE REVIEW**

There is an issue in the manager start up due a lack of sync between the certificate creation and the webhook server trying to read it. I'm working on this and will update the PR soon. controller-runtime seems to have a module which can help on that and I'm validating it. But I want to give you the opportunity to read the first version of the code to speed up the review process. 

The package or their names can be rearranged in the future. For example, the `admissionregistration` package could be rewritten soon. 

## Test

You will not be able to run this in host machine. The code that setup the certificate locally in the host has not been updated yet. You need to build the container, load it in a cluster and install Kubewarden using helm charts.

### CODE REVIEW

**Comment by jvanz** [comment]
@kubewarden/kubewarden-developers the controller is changed to create the root CA, the certificates for the webhook server and update the webook to use the root CA. 

To prevent some concurrency issues like trying to read the certificate file before its available a custom webhook server has been created to wait for  the job that create/check the certificates. Furthermore,  to prevent the control plane to contact the server before the certificates the ready checker return only after the certificates are ready. 

Keep in mind that routine can change when implementing the certificate rotation. And file structure can be updated as well. 

To properly test this change you can build the container image and test with this [branch](https://github.com/kubewarden/helm-charts/tree/ca-changes) of the helm chart. This is a command that you can use:

```
make docker-build  IMG=ghcr.io/jvanz/kubewarden-controller:root-ca-change && 
pushd . &&
cd <path to the kubewarden e2e tests repository> && 
make clean cluster && 
k3d image import ghcr.io/jvanz/kubewarden-controller:root-ca-change --cluster kubewarden-testing && 
make install KUBEWARDEN_CHARTS_LOCATION=<path to helm chart repository dir with the ca-changes branch code>
popd
```

File: internal/pkg/crypto/ca.go
Comment (viccuad) [inline internal/pkg/crypto/ca.go:21]: I think we should keep returning encoded PEM as strings, given that in it we get base64 payload, ascii compatible, so utf compatible. Hence why they have been encoded.

File: internal/pkg/constants/constants.go
Comment (viccuad) [inline internal/pkg/constants/constants.go:12]: Silly nitpick with CamelCase:

```suggestion
	ControllerCertsSecretName = "kubewarden-controller-certificate"
```

File: internal/pkg/certificates/certificates.go
Comment (viccuad) [inline internal/pkg/certificates/certificates.go:67]: I would add a comment explaining this;
"We have updated the certs and the webhooks, trigger a new controller container instance that will load the new and correct ones, by closing the current one with os.Exit"
or something similar

File: internal/pkg/certificates/certificates.go
Comment (viccuad) [inline internal/pkg/certificates/certificates.go:91]: do we need to allocate memory for this bool `false`?

**Review by viccuad — COMMENTED** [review]
It looks good! :heart: . I have some questions and simple nitpicks before approving.

File: internal/pkg/certificates/certificates.go
Comment (flavio) [inline internal/pkg/certificates/certificates.go:23]:
```suggestion
// controller. If the certificate is created, the process exit to force kubernetes to redeploy
```

File: internal/pkg/certificates/certificates.go
Comment (flavio) [inline internal/pkg/certificates/certificates.go:26]: This should be a debug statement, plus the message is a bit misleading. We are verifying the root CA, not the certificates

File: internal/pkg/certificates/certificates.go
Comment (flavio) [inline internal/pkg/certificates/certificates.go:31]: I would move all this inner block to a dedicated function, like `createRootCA` that would generate the root CA and write that into Kubernetes.

Out of memory, I think we used to have this function

File: internal/pkg/certificates/certificates.go
Comment (flavio) [reply internal/pkg/certificates/certificates.go:67]: I would add a log statement that explains why the controller is exiting

File: internal/pkg/certificates/certificates.go
Comment (flavio) [inline internal/pkg/certificates/certificates.go:75]: I think this is misleading. From what I understand this is executed only if the secret that contains the root CA has been found. No verification has been done. I think we can omit this check

File: internal/pkg/certificates/certificates.go
Comment (flavio) [inline internal/pkg/certificates/certificates.go:25]: Having reviewed the code, this function makes sure the root CA is defined inside of a specified secret. If the root CA is not found, then the controller creates it, stores it into Kubernetes and then creates a new certificate to be used by the webhook endpoint provided by the controller.

I think the name of the function is misleading

File: internal/pkg/certificates/certificates.go
Comment (flavio) [reply internal/pkg/certificates/certificates.go:91]: I **think** this is done because `Immutable` is a `*bool`, hence we either declare a variable before and then reference it, or we use `new`.

The end result is the same, isn't it?

File: internal/pkg/crypto/ca.go
Comment (flavio) [inline internal/pkg/crypto/ca.go:33]: Same here, is there a reason to not return a `string`?

File: internal/pkg/crypto/integration_test.go
Comment (flavio) [inline internal/pkg/crypto/integration_test.go:15]: If we want to make sure the certificate being issued is valid, we could use the `Certificate.Verify` function, like done inside of the [example](https://pkg.go.dev/crypto/x509#Certificate.Verify) provided by the std library.
Using a http server is overly complicated IMHO.

**Review by flavio — CHANGES_REQUESTED** [review]

File: .gitignore
Comment (fabriziosestito) [inline .gitignore:29]: why is this needed?

File: internal/pkg/certificates/certificates.go
Comment (viccuad) [reply internal/pkg/certificates/certificates.go:91]: I think you are right, in that case, ignore this conversation.

File: .gitignore
Comment (jvanz) [reply .gitignore:29]: Just to avoid adding the vendor directory in a commit.

File: internal/pkg/certificates/certificates.go
Comment (jvanz) [reply internal/pkg/certificates/certificates.go:32]: I think you are talking about the function inside the `internal/pkg/admission` and `internal/pkg/admissionregistration`. I'm aware of that. Actually, the `internal/pkg/certificates` package is a kind of copy of some of those functions. However, I have not merge all of the code yet. Just to avoid having a huge PR as before. I'll merge the code  when I start to change the reconciliation loop for policy server and policies. For now, I'm reorganizing the necessary  functions on demand.

File: internal/pkg/crypto/ca.go
Comment (jvanz) [reply internal/pkg/crypto/ca.go:21]: To void another conversion. This function is returning `*bytes.Buffer` because I do not need string value. All the usage of the data returned by this function needs to be bytes. Not strings. But, I keep using `*bytes.Buffer` instead of `[]byte` because if in the next changes is necessary a string I could access it using the available function [`String`](https://pkg.go.dev/bytes@go1.21.4#Buffer.String) or [`Bytes`](https://pkg.go.dev/bytes@go1.21.4#Buffer.Bytes).

File: internal/pkg/crypto/ca.go
Comment (jvanz) [reply internal/pkg/crypto/ca.go:33]: To void another conversion. This function is returning `*bytes.Buffer` because I do not need string value. All the usage of the data returned by this function needs to be bytes. Not strings. But, I keep using `*bytes.Buffer` instead of `[]byte` because if in the next changes is necessary a string I could access it using the available function [`String`](https://pkg.go.dev/bytes@go1.21.4#Buffer.String) or [`Bytes`](https://pkg.go.dev/bytes@go1.21.4#Buffer.Bytes).

File: internal/pkg/crypto/integration_test.go
Comment (jvanz) [reply internal/pkg/crypto/integration_test.go:15]: I see. Indeed, that makes sense. As I said in a previous comment, as this package is a children from other packages in the `internal/pkg` directory. Which includes the test cases as well. Let me adapt the tests

File: internal/pkg/admission/policy-server-ca-secret_test.go
Comment (flavio) [inline internal/pkg/admission/policy-server-ca-secret_test.go:35]: Is there something that prevents us from having this test case enabled as part of this PR?

File: internal/pkg/admission/policy-server-ca-secret_test.go
Comment (flavio) [inline internal/pkg/admission/policy-server-ca-secret_test.go:41]: I think we should add an assertion that ensures `err` is `nil`

File: internal/pkg/admission/policy-server-ca-secret_test.go
Comment (flavio) [inline internal/pkg/admission/policy-server-ca-secret_test.go:71]: same here

File: internal/pkg/admission/policy-server-ca-secret_test.go
Comment (flavio) [inline internal/pkg/admission/policy-server-ca-secret_test.go:77]: same here

File: internal/pkg/certificates/certificates.go
Comment (flavio) [inline internal/pkg/certificates/certificates.go:38]: nitpicking, maybe we should create some variables to hold the results of the `fmt.Sprintf`. The names of these variables would make clear what they are about to someone reading the code

File: internal/pkg/certificates/certificates.go
Comment (flavio) [inline internal/pkg/certificates/certificates.go:54]: We can leave that out (and update the function documentation accordingly):

* We are going to create the certificate before starting the webhook server
* The webhook server provided by controller runtime can detect changes done to the certificate & its key. We just need to update the Secret, kubelet will sync the contents on the files mounted inside of the Pod, controller-runtime will reload them and start answering with the new cert

File: internal/pkg/certificates/certificates.go
Comment (flavio) [inline internal/pkg/certificates/certificates.go:50]: Can you please switch to the more idiomatic:

```golang
if err = foo(); err != nil {
}
```

File: internal/pkg/certificates/certificates.go
Comment (flavio) [inline internal/pkg/certificates/certificates.go:84]: Can you please switch to the more idiomatic:

```golang
if err = foo(); err != nil {
}
```

File: internal/pkg/certificates/certificates.go
Comment (flavio) [inline internal/pkg/certificates/certificates.go:117]: The controller's webhook certs have a hard coded name, we should just do a `Get`

File: internal/pkg/certificates/certificates.go
Comment (flavio) [inline internal/pkg/certificates/certificates.go:133]: The controller's webhook certs have a hard coded name, we should just do a `Get`

File: internal/pkg/constants/constants.go
Comment (flavio) [inline internal/pkg/constants/constants.go:62]: I think we can remove this one, see my previous comment about exiting once the CA is updated/initialized

File: internal/pkg/crypto/cert.go
Comment (flavio) [inline internal/pkg/crypto/cert.go:52]: Let's do it right now. It looks like we are not setting the SAN

File: internal/pkg/crypto/integration_test.go
Comment (flavio) [reply internal/pkg/crypto/integration_test.go:15]: Reminder to update the test case

File: internal/pkg/webhookserver/server.go
Comment (flavio) [inline internal/pkg/webhookserver/server.go:1]: Let's either move the the synchronous init as part of this PR or let's do that with another PR. In both cases, let's remove this file (if I got this right, this is required to have the async init)

File: internal/pkg/constants/constants.go
Comment (flavio) [reply internal/pkg/constants/constants.go:62]: On top of that, if we remove the exit status from above I think we can remove all the other constants and just revert to the initial code.

File: main.go
Comment (flavio) [inline main.go:100]: Do we really need it? I think this was needed in the past before we used Tilt for local development.

File: main.go
Comment (flavio) [inline main.go:67]: I think we can revert this change, see my previous comment about the exit codes

File: main.go
Comment (flavio) [inline main.go:113]: I think we can revert this change, see my previous comment about the exit codes

File: main.go
Comment (flavio) [inline main.go:126]: I think we can revert this change, see my previous comment about the exit codes

**Review by flavio — CHANGES_REQUESTED** [review]
Good improvement over the previous iteration. Thanks for having removed the code duplication!

File: internal/pkg/webhookserver/server.go
Comment (fabriziosestito) [reply internal/pkg/webhookserver/server.go:1]: I think that this is not needed anymore

**Comment by jvanz** [comment]
@kubewarden/kubewarden-developers this is not ready for review. I've moved it back to in progress. It still missing the sync cert creation.

**Comment by jvanz** [comment]
@kubewarden/kubewarden-developers can you do another round of reviews? I still want to do some tests. But consider the time and timezone differences, I'll not finish Today and I would like to work in your possible feedback Tomorrow. ;-)

File: internal/pkg/crypto/ca.go
Comment (jvanz) [inline internal/pkg/crypto/ca.go:133]: I'm aware that this function is very similar to another in the `internal/pkg/admission/policy-server-ca-secret.go`. I have not remove that function because that require further changes in the reconciliation loop to keep  the tests passing and I would like to keep this PR as small as possible. This reconciliation loop changes will be the next step.

File: main.go
Comment (viccuad) [inline main.go:205]: question: this is for injecting the kubebuilder scaffolded code for the controllerManager that does the `setupWithManager()`, correct?
I couldn't find any docs, only this [comment](https://togithub.com/kubernetes-sigs/kubebuilder/issues/1487#issuecomment-1485346413). The relevant code seems to be [pkg/machinery](https://github.com/kubernetes-sigs/kubebuilder/blob/a897dae204d9b32483d14c60a78573fca6827378/pkg/machinery).

File: main.go
Comment (viccuad) [inline main.go:13]: whoops, a word was eaten.

File: internal/pkg/crypto/ca.go
Comment (viccuad) [reply internal/pkg/crypto/ca.go:24]: I still think that String would be a better fit than bytes.Buffer or []Bytes. Once we have created the PEM encoded certificate, we don't need to change it (String is immutable, the other not), we don't need to sequentially add chars (so we don't need a Buffer that is cost-effective for that), and the result is supposed to be human-readable (String).

File: internal/pkg/certificates/certificates.go
Comment (viccuad) [inline internal/pkg/certificates/certificates.go:77]: I like this, just to make sure that we have renewed correctly.

File: internal/pkg/admission/policy-server-ca-secret.go
Comment (viccuad) [inline internal/pkg/admission/policy-server-ca-secret.go:90]: Nit for clarity:

```suggestion
	if !hasCARootCertKey {
```

**Review by viccuad — COMMENTED** [review]
Looks good here! I would like the PEM encoded certs to be stored as String though, but that's it.

**Comment by jvanz** [comment]
> Looks good here! I would like the PEM encoded certs to be stored as String though, but that's it.

Do you mean the data in the secret?

**Comment by viccuad** [comment]
> @viccuad , do you mean the data in the secret?

No, I was refering to https://github.com/kubewarden/kubewarden-controller/pull/550#discussion_r1427862729

File: internal/pkg/certificates/certificates.go
Comment (fabriziosestito) [inline internal/pkg/certificates/certificates.go:225]: nit(readability): we could move the public function on top of the file so it's easier to understand the public interfaces.

File: internal/pkg/admission/reconciler.go
Comment (fabriziosestito) [inline internal/pkg/admission/reconciler.go:157]: nice one, I like we are getting read of the inversion of control here

File: Tiltfile
Comment (fabriziosestito) [inline Tiltfile:13]: we can remove this

File: internal/pkg/crypto/ca.go
Comment (fabriziosestito) [inline internal/pkg/crypto/ca.go:1]: I think that the  package  name is a bit confusing. Usually crypto is used for packages that deal with cryptography algorithms, in this case I think this package can be merged with the certificates package

**Review by fabriziosestito — APPROVED** [review]
Overall LGTM!
Just some minor comments added.

I see that the policy-server controller now uses the certificate generation from the crypto package. Ideally we should use the same certificate generated in the main setup function. Is this planned for the next iteration?

File: internal/pkg/crypto/ca.go
Comment (jvanz) [reply internal/pkg/crypto/ca.go:1]: That's right. I'll push the changes fixing that soon

**Review by viccuad — APPROVED** [review]
LGTM! Quite a job!

The only thing I don't like is https://github.com/kubewarden/kubewarden-controller/pull/550#discussion_r1360424422, but I don't want to block the PR for it.

**Pull request merged**
